In [2]:
import pandas as pd
import osmapi
import time

In [3]:
api = osmapi.OsmApi()

In [4]:
df = pd.read_csv("data/dataviz.csv")
print(df)

           plat      plng       dlat      dlng         t  \
0     46.517116  6.630342  46.508528  6.627598  11:15:00   
1     46.517116  6.630342  46.521429  6.656347  20:00:00   
2     46.517116  6.630342  46.558792  6.680104  18:15:00   
3     46.520649  6.635173  46.519590  6.634718  18:45:00   
4     46.510731  6.630553  46.515950  6.657525  19:00:00   
5     46.520649  6.635173  46.558792  6.680104  18:30:00   
6     46.517116  6.630342  46.537950  6.571898  18:00:00   
7     46.520649  6.635173  46.558792  6.680104  20:15:00   
8     46.518914  6.637912  46.537411  6.653249  19:30:00   
9     46.517116  6.630342  46.524409  6.629936  19:15:00   
10    46.520649  6.635173  46.530030  6.572823  19:45:00   
11    46.527252  6.623526  46.558792  6.680104  21:00:00   
12    46.517116  6.630342  46.509917  6.622856  18:15:00   
13    46.520649  6.635173  46.541796  6.623339  11:15:00   
14    46.517116  6.630342  46.515180  6.633784  18:45:00   
15    46.520649  6.635173  46.526995  6.

In [5]:
#pickup_points = df[['plat', 'plng']].drop_duplicates().reset_index(drop = True)

In [5]:
unique_nodes = set()
for ids in df['road']:
    id_list = ids.split(', ')
    id_list[0] = id_list[0][1:]
    id_list[-1] = id_list[-1][:-1]
    for i in id_list:
        if i not in unique_nodes:
            unique_nodes.add(i)
#print(unique_nodes)
print(len(unique_nodes))

12600


In [59]:
#unique_nodes = {2223091605, 311463563}

In [17]:
f = open("node_id_to_coordinates.csv","w+")
f.write('node_id,lat,lon\n')

total = len(unique_nodes)
count = 1
error_count = 0
for node_id in unique_nodes:
    print(str(count) + '/' + str(total))
    count += 1
    time.sleep(0.1)
    try:
        node = api.NodeGet(node_id)
        f.write(str(node_id) + ', ' + str(node['lat']) + ', ' + str(node['lon']) + '\n')
    except:
        print('ERROR ON ' + str(node_id))
        error_count += 1
 
f.close()


1/12600
2/12600
3/12600
4/12600
5/12600
6/12600
7/12600
8/12600
9/12600
10/12600
11/12600
12/12600
13/12600
14/12600
15/12600
16/12600
17/12600
18/12600
19/12600
20/12600
21/12600
22/12600
23/12600
24/12600
25/12600
26/12600
27/12600
28/12600
29/12600
30/12600
31/12600
32/12600
33/12600
34/12600
35/12600
36/12600
37/12600
38/12600
39/12600
40/12600
41/12600
42/12600
43/12600
44/12600
45/12600
46/12600
47/12600
48/12600
49/12600
50/12600
51/12600
52/12600
53/12600
54/12600
55/12600
56/12600
57/12600
58/12600
59/12600
60/12600
61/12600
62/12600
63/12600
64/12600
65/12600
66/12600
67/12600
68/12600
69/12600
70/12600
71/12600
72/12600
73/12600
74/12600
75/12600
76/12600
77/12600
78/12600
79/12600
80/12600
81/12600
82/12600
83/12600
84/12600
85/12600
86/12600
87/12600
88/12600
89/12600
90/12600
91/12600
92/12600
93/12600
94/12600
95/12600
96/12600
97/12600
98/12600
99/12600
100/12600
101/12600
102/12600
103/12600
104/12600
105/12600
106/12600
107/12600
108/12600
109/12600
110/12600
111/1260

830/12600
831/12600
832/12600
833/12600
834/12600
835/12600
836/12600
837/12600
838/12600
839/12600
840/12600
841/12600
842/12600
843/12600
844/12600
845/12600
846/12600
847/12600
848/12600
849/12600
850/12600
851/12600
852/12600
853/12600
854/12600
855/12600
856/12600
857/12600
858/12600
859/12600
860/12600
861/12600
862/12600
863/12600
864/12600
865/12600
866/12600
867/12600
868/12600
869/12600
870/12600
871/12600
872/12600
873/12600
874/12600
875/12600
876/12600
877/12600
878/12600
879/12600
880/12600
881/12600
882/12600
883/12600
884/12600
885/12600
886/12600
887/12600
888/12600
889/12600
890/12600
891/12600
892/12600
893/12600
894/12600
895/12600
896/12600
897/12600
898/12600
899/12600
900/12600
901/12600
902/12600
903/12600
904/12600
905/12600
906/12600
907/12600
908/12600
909/12600
910/12600
911/12600
912/12600
913/12600
914/12600
915/12600
916/12600
917/12600
918/12600
919/12600
920/12600
921/12600
922/12600
923/12600
924/12600
925/12600
926/12600
927/12600
928/12600
929/12600


1589/12600
1590/12600
1591/12600
1592/12600
1593/12600
1594/12600
1595/12600
1596/12600
1597/12600
1598/12600
1599/12600
1600/12600
1601/12600
1602/12600
1603/12600
1604/12600
1605/12600
1606/12600
1607/12600
1608/12600
1609/12600
1610/12600
1611/12600
1612/12600
1613/12600
1614/12600
1615/12600
1616/12600
1617/12600
1618/12600
1619/12600
1620/12600
1621/12600
1622/12600
1623/12600
1624/12600
1625/12600
1626/12600
1627/12600
1628/12600
1629/12600
1630/12600
1631/12600
1632/12600
1633/12600
1634/12600
1635/12600
1636/12600
1637/12600
1638/12600
1639/12600
1640/12600
1641/12600
1642/12600
1643/12600
1644/12600
1645/12600
1646/12600
1647/12600
1648/12600
1649/12600
1650/12600
1651/12600
1652/12600
1653/12600
1654/12600
1655/12600
1656/12600
1657/12600
1658/12600
1659/12600
1660/12600
1661/12600
1662/12600
1663/12600
1664/12600
1665/12600
1666/12600
1667/12600
1668/12600
1669/12600
1670/12600
1671/12600
1672/12600
1673/12600
1674/12600
1675/12600
1676/12600
1677/12600
1678/12600
1679/12600

2334/12600
2335/12600
2336/12600
2337/12600
2338/12600
2339/12600
2340/12600
2341/12600
2342/12600
2343/12600
2344/12600
2345/12600
2346/12600
2347/12600
2348/12600
2349/12600
2350/12600
2351/12600
2352/12600
2353/12600
2354/12600
2355/12600
2356/12600
2357/12600
2358/12600
2359/12600
2360/12600
2361/12600
2362/12600
2363/12600
2364/12600
2365/12600
2366/12600
2367/12600
2368/12600
2369/12600
2370/12600
2371/12600
2372/12600
2373/12600
2374/12600
2375/12600
2376/12600
2377/12600
2378/12600
2379/12600
2380/12600
2381/12600
2382/12600
2383/12600
2384/12600
2385/12600
2386/12600
2387/12600
2388/12600
2389/12600
2390/12600
2391/12600
2392/12600
2393/12600
2394/12600
2395/12600
2396/12600
2397/12600
2398/12600
2399/12600
2400/12600
2401/12600
2402/12600
2403/12600
2404/12600
2405/12600
2406/12600
2407/12600
2408/12600
2409/12600
2410/12600
2411/12600
2412/12600
2413/12600
2414/12600
2415/12600
2416/12600
2417/12600
2418/12600
2419/12600
2420/12600
2421/12600
2422/12600
2423/12600
2424/12600

3078/12600
3079/12600
3080/12600
3081/12600
3082/12600
3083/12600
3084/12600
3085/12600
3086/12600
3087/12600
3088/12600
3089/12600
3090/12600
3091/12600
3092/12600
3093/12600
3094/12600
3095/12600
3096/12600
3097/12600
3098/12600
3099/12600
3100/12600
3101/12600
3102/12600
3103/12600
3104/12600
3105/12600
3106/12600
3107/12600
3108/12600
3109/12600
3110/12600
3111/12600
3112/12600
3113/12600
3114/12600
3115/12600
3116/12600
3117/12600
3118/12600
3119/12600
3120/12600
3121/12600
3122/12600
3123/12600
3124/12600
3125/12600
3126/12600
3127/12600
3128/12600
3129/12600
3130/12600
3131/12600
3132/12600
3133/12600
3134/12600
3135/12600
3136/12600
3137/12600
3138/12600
3139/12600
3140/12600
ERROR ON 582463773
3141/12600
3142/12600
3143/12600
3144/12600
3145/12600
3146/12600
3147/12600
3148/12600
3149/12600
3150/12600
3151/12600
3152/12600
3153/12600
3154/12600
3155/12600
3156/12600
3157/12600
3158/12600
3159/12600
3160/12600
3161/12600
3162/12600
3163/12600
3164/12600
3165/12600
3166/12600
31

3818/12600
3819/12600
3820/12600
3821/12600
3822/12600
3823/12600
3824/12600
3825/12600
3826/12600
3827/12600
3828/12600
3829/12600
3830/12600
3831/12600
3832/12600
3833/12600
ERROR ON 580047011
3834/12600
3835/12600
3836/12600
3837/12600
3838/12600
3839/12600
3840/12600
3841/12600
3842/12600
3843/12600
3844/12600
3845/12600
3846/12600
3847/12600
3848/12600
3849/12600
3850/12600
3851/12600
3852/12600
3853/12600
3854/12600
3855/12600
3856/12600
3857/12600
3858/12600
3859/12600
3860/12600
3861/12600
3862/12600
3863/12600
3864/12600
3865/12600
3866/12600
3867/12600
3868/12600
3869/12600
3870/12600
3871/12600
3872/12600
3873/12600
3874/12600
3875/12600
3876/12600
3877/12600
3878/12600
3879/12600
3880/12600
3881/12600
3882/12600
3883/12600
3884/12600
3885/12600
3886/12600
3887/12600
3888/12600
3889/12600
3890/12600
3891/12600
3892/12600
3893/12600
3894/12600
3895/12600
3896/12600
3897/12600
3898/12600
3899/12600
3900/12600
3901/12600
3902/12600
3903/12600
3904/12600
3905/12600
3906/12600
39

4563/12600
4564/12600
4565/12600
4566/12600
4567/12600
4568/12600
4569/12600
4570/12600
4571/12600
4572/12600
4573/12600
4574/12600
4575/12600
4576/12600
4577/12600
4578/12600
4579/12600
4580/12600
4581/12600
4582/12600
4583/12600
4584/12600
4585/12600
4586/12600
4587/12600
4588/12600
4589/12600
4590/12600
4591/12600
4592/12600
4593/12600
4594/12600
4595/12600
4596/12600
4597/12600
4598/12600
4599/12600
4600/12600
4601/12600
4602/12600
4603/12600
4604/12600
4605/12600
4606/12600
4607/12600
4608/12600
4609/12600
4610/12600
4611/12600
4612/12600
4613/12600
4614/12600
4615/12600
4616/12600
4617/12600
4618/12600
4619/12600
4620/12600
4621/12600
4622/12600
4623/12600
4624/12600
ERROR ON 376010056
4625/12600
4626/12600
4627/12600
4628/12600
4629/12600
4630/12600
4631/12600
4632/12600
4633/12600
4634/12600
4635/12600
4636/12600
4637/12600
4638/12600
4639/12600
4640/12600
4641/12600
4642/12600
4643/12600
4644/12600
4645/12600
4646/12600
4647/12600
4648/12600
4649/12600
4650/12600
4651/12600
46

5308/12600
5309/12600
5310/12600
5311/12600
5312/12600
5313/12600
5314/12600
5315/12600
5316/12600
5317/12600
5318/12600
5319/12600
5320/12600
5321/12600
5322/12600
5323/12600
5324/12600
5325/12600
5326/12600
5327/12600
5328/12600
5329/12600
5330/12600
5331/12600
5332/12600
5333/12600
5334/12600
5335/12600
5336/12600
5337/12600
5338/12600
5339/12600
5340/12600
5341/12600
5342/12600
5343/12600
5344/12600
5345/12600
5346/12600
5347/12600
5348/12600
5349/12600
5350/12600
5351/12600
5352/12600
5353/12600
5354/12600
5355/12600
5356/12600
5357/12600
5358/12600
5359/12600
5360/12600
5361/12600
5362/12600
5363/12600
5364/12600
5365/12600
5366/12600
5367/12600
5368/12600
5369/12600
5370/12600
5371/12600
5372/12600
5373/12600
5374/12600
5375/12600
5376/12600
5377/12600
5378/12600
5379/12600
5380/12600
5381/12600
5382/12600
5383/12600
5384/12600
5385/12600
5386/12600
5387/12600
5388/12600
5389/12600
5390/12600
5391/12600
5392/12600
5393/12600
5394/12600
5395/12600
5396/12600
5397/12600
5398/12600

6053/12600
6054/12600
6055/12600
6056/12600
6057/12600
6058/12600
6059/12600
6060/12600
6061/12600
6062/12600
6063/12600
6064/12600
6065/12600
6066/12600
6067/12600
6068/12600
6069/12600
6070/12600
6071/12600
ERROR ON 312416320
6072/12600
6073/12600
6074/12600
6075/12600
6076/12600
6077/12600
6078/12600
6079/12600
6080/12600
6081/12600
6082/12600
6083/12600
6084/12600
6085/12600
6086/12600
6087/12600
6088/12600
6089/12600
6090/12600
6091/12600
6092/12600
6093/12600
6094/12600
6095/12600
6096/12600
6097/12600
6098/12600
6099/12600
6100/12600
6101/12600
6102/12600
6103/12600
6104/12600
6105/12600
6106/12600
6107/12600
6108/12600
6109/12600
6110/12600
6111/12600
6112/12600
6113/12600
6114/12600
6115/12600
6116/12600
6117/12600
6118/12600
6119/12600
6120/12600
6121/12600
6122/12600
6123/12600
6124/12600
6125/12600
6126/12600
6127/12600
6128/12600
6129/12600
6130/12600
6131/12600
6132/12600
6133/12600
6134/12600
6135/12600
6136/12600
6137/12600
6138/12600
6139/12600
6140/12600
6141/12600
61

6795/12600
6796/12600
6797/12600
6798/12600
6799/12600
6800/12600
6801/12600
6802/12600
6803/12600
6804/12600
6805/12600
6806/12600
6807/12600
6808/12600
6809/12600
6810/12600
6811/12600
6812/12600
6813/12600
6814/12600
6815/12600
6816/12600
6817/12600
6818/12600
6819/12600
6820/12600
6821/12600
6822/12600
6823/12600
6824/12600
6825/12600
6826/12600
6827/12600
6828/12600
6829/12600
6830/12600
6831/12600
6832/12600
6833/12600
6834/12600
6835/12600
6836/12600
6837/12600
6838/12600
6839/12600
6840/12600
6841/12600
6842/12600
6843/12600
6844/12600
6845/12600
6846/12600
6847/12600
6848/12600
6849/12600
6850/12600
6851/12600
6852/12600
6853/12600
6854/12600
6855/12600
6856/12600
6857/12600
6858/12600
6859/12600
6860/12600
6861/12600
6862/12600
6863/12600
6864/12600
6865/12600
6866/12600
6867/12600
6868/12600
6869/12600
6870/12600
6871/12600
6872/12600
6873/12600
6874/12600
6875/12600
6876/12600
6877/12600
6878/12600
6879/12600
6880/12600
6881/12600
6882/12600
6883/12600
6884/12600
6885/12600

7538/12600
7539/12600
7540/12600
7541/12600
7542/12600
7543/12600
7544/12600
7545/12600
7546/12600
7547/12600
7548/12600
7549/12600
7550/12600
7551/12600
7552/12600
7553/12600
7554/12600
7555/12600
7556/12600
7557/12600
7558/12600
7559/12600
7560/12600
7561/12600
7562/12600
7563/12600
7564/12600
7565/12600
7566/12600
7567/12600
7568/12600
7569/12600
7570/12600
7571/12600
7572/12600
7573/12600
7574/12600
7575/12600
7576/12600
7577/12600
7578/12600
7579/12600
7580/12600
7581/12600
7582/12600
7583/12600
7584/12600
7585/12600
7586/12600
7587/12600
7588/12600
7589/12600
7590/12600
7591/12600
7592/12600
7593/12600
7594/12600
7595/12600
7596/12600
7597/12600
7598/12600
7599/12600
7600/12600
7601/12600
7602/12600
7603/12600
7604/12600
7605/12600
7606/12600
7607/12600
7608/12600
7609/12600
7610/12600
7611/12600
7612/12600
7613/12600
7614/12600
7615/12600
7616/12600
7617/12600
7618/12600
7619/12600
7620/12600
7621/12600
7622/12600
7623/12600
7624/12600
7625/12600
7626/12600
7627/12600
7628/12600

8283/12600
8284/12600
8285/12600
8286/12600
8287/12600
8288/12600
8289/12600
8290/12600
8291/12600
8292/12600
8293/12600
8294/12600
8295/12600
8296/12600
8297/12600
8298/12600
8299/12600
8300/12600
8301/12600
8302/12600
8303/12600
8304/12600
8305/12600
8306/12600
8307/12600
8308/12600
8309/12600
8310/12600
8311/12600
8312/12600
8313/12600
8314/12600
8315/12600
8316/12600
8317/12600
8318/12600
8319/12600
8320/12600
8321/12600
8322/12600
8323/12600
8324/12600
8325/12600
8326/12600
8327/12600
8328/12600
8329/12600
8330/12600
8331/12600
8332/12600
8333/12600
8334/12600
8335/12600
8336/12600
8337/12600
8338/12600
8339/12600
8340/12600
8341/12600
8342/12600
8343/12600
8344/12600
8345/12600
8346/12600
8347/12600
8348/12600
8349/12600
8350/12600
8351/12600
8352/12600
8353/12600
8354/12600
8355/12600
8356/12600
8357/12600
8358/12600
8359/12600
8360/12600
8361/12600
8362/12600
8363/12600
8364/12600
ERROR ON 660206860
8365/12600
8366/12600
8367/12600
8368/12600
8369/12600
8370/12600
8371/12600
83

9028/12600
9029/12600
9030/12600
9031/12600
9032/12600
9033/12600
9034/12600
9035/12600
9036/12600
9037/12600
9038/12600
9039/12600
9040/12600
9041/12600
9042/12600
9043/12600
9044/12600
9045/12600
9046/12600
9047/12600
9048/12600
9049/12600
9050/12600
9051/12600
9052/12600
9053/12600
9054/12600
9055/12600
9056/12600
9057/12600
9058/12600
9059/12600
9060/12600
9061/12600
9062/12600
9063/12600
9064/12600
9065/12600
9066/12600
9067/12600
9068/12600
9069/12600
9070/12600
9071/12600
9072/12600
9073/12600
9074/12600
9075/12600
9076/12600
9077/12600
9078/12600
9079/12600
9080/12600
9081/12600
9082/12600
9083/12600
9084/12600
9085/12600
9086/12600
9087/12600
9088/12600
9089/12600
9090/12600
9091/12600
9092/12600
9093/12600
9094/12600
9095/12600
9096/12600
9097/12600
9098/12600
9099/12600
9100/12600
9101/12600
9102/12600
9103/12600
9104/12600
9105/12600
9106/12600
9107/12600
9108/12600
9109/12600
9110/12600
9111/12600
9112/12600
9113/12600
9114/12600
9115/12600
9116/12600
9117/12600
9118/12600

9771/12600
9772/12600
9773/12600
9774/12600
9775/12600
9776/12600
9777/12600
9778/12600
9779/12600
9780/12600
9781/12600
9782/12600
9783/12600
9784/12600
9785/12600
9786/12600
9787/12600
9788/12600
9789/12600
9790/12600
9791/12600
9792/12600
9793/12600
9794/12600
9795/12600
9796/12600
9797/12600
9798/12600
9799/12600
9800/12600
9801/12600
9802/12600
9803/12600
9804/12600
9805/12600
9806/12600
9807/12600
9808/12600
9809/12600
9810/12600
9811/12600
9812/12600
9813/12600
9814/12600
9815/12600
9816/12600
9817/12600
9818/12600
9819/12600
9820/12600
9821/12600
9822/12600
9823/12600
9824/12600
9825/12600
9826/12600
9827/12600
9828/12600
9829/12600
9830/12600
9831/12600
9832/12600
9833/12600
9834/12600
9835/12600
9836/12600
9837/12600
9838/12600
9839/12600
9840/12600
9841/12600
9842/12600
9843/12600
9844/12600
9845/12600
9846/12600
9847/12600
9848/12600
9849/12600
9850/12600
9851/12600
9852/12600
9853/12600
9854/12600
9855/12600
9856/12600
9857/12600
9858/12600
9859/12600
9860/12600
9861/12600

10472/12600
10473/12600
10474/12600
10475/12600
10476/12600
10477/12600
10478/12600
10479/12600
10480/12600
10481/12600
10482/12600
10483/12600
10484/12600
10485/12600
10486/12600
10487/12600
10488/12600
10489/12600
10490/12600
10491/12600
10492/12600
10493/12600
10494/12600
10495/12600
10496/12600
10497/12600
10498/12600
10499/12600
10500/12600
10501/12600
10502/12600
10503/12600
10504/12600
10505/12600
10506/12600
10507/12600
10508/12600
10509/12600
10510/12600
10511/12600
10512/12600
10513/12600
10514/12600
10515/12600
10516/12600
10517/12600
10518/12600
10519/12600
10520/12600
10521/12600
10522/12600
10523/12600
10524/12600
10525/12600
10526/12600
10527/12600
10528/12600
10529/12600
10530/12600
10531/12600
10532/12600
10533/12600
10534/12600
10535/12600
10536/12600
10537/12600
10538/12600
10539/12600
10540/12600
10541/12600
10542/12600
10543/12600
10544/12600
10545/12600
10546/12600
10547/12600
10548/12600
10549/12600
10550/12600
10551/12600
10552/12600
10553/12600
10554/12600
1055

11155/12600
11156/12600
11157/12600
11158/12600
11159/12600
11160/12600
11161/12600
11162/12600
11163/12600
11164/12600
11165/12600
11166/12600
11167/12600
11168/12600
11169/12600
11170/12600
11171/12600
11172/12600
11173/12600
11174/12600
11175/12600
11176/12600
11177/12600
11178/12600
11179/12600
11180/12600
11181/12600
11182/12600
11183/12600
11184/12600
11185/12600
11186/12600
11187/12600
11188/12600
11189/12600
11190/12600
11191/12600
11192/12600
11193/12600
11194/12600
11195/12600
11196/12600
11197/12600
11198/12600
11199/12600
11200/12600
11201/12600
11202/12600
11203/12600
11204/12600
11205/12600
11206/12600
11207/12600
11208/12600
11209/12600
11210/12600
11211/12600
11212/12600
11213/12600
11214/12600
11215/12600
11216/12600
11217/12600
11218/12600
11219/12600
11220/12600
11221/12600
11222/12600
11223/12600
11224/12600
11225/12600
11226/12600
11227/12600
11228/12600
11229/12600
11230/12600
11231/12600
11232/12600
11233/12600
11234/12600
11235/12600
11236/12600
11237/12600
1123

11839/12600
11840/12600
11841/12600
11842/12600
11843/12600
11844/12600
11845/12600
11846/12600
11847/12600
11848/12600
11849/12600
11850/12600
11851/12600
11852/12600
11853/12600
11854/12600
11855/12600
11856/12600
11857/12600
11858/12600
11859/12600
11860/12600
11861/12600
11862/12600
11863/12600
11864/12600
11865/12600
11866/12600
11867/12600
11868/12600
11869/12600
11870/12600
11871/12600
11872/12600
11873/12600
11874/12600
11875/12600
11876/12600
11877/12600
11878/12600
11879/12600
11880/12600
11881/12600
11882/12600
11883/12600
11884/12600
11885/12600
11886/12600
11887/12600
11888/12600
11889/12600
11890/12600
11891/12600
11892/12600
11893/12600
11894/12600
11895/12600
11896/12600
11897/12600
11898/12600
11899/12600
11900/12600
11901/12600
11902/12600
11903/12600
11904/12600
11905/12600
11906/12600
11907/12600
11908/12600
11909/12600
11910/12600
11911/12600
11912/12600
11913/12600
11914/12600
11915/12600
11916/12600
11917/12600
11918/12600
11919/12600
11920/12600
11921/12600
1192

12522/12600
12523/12600
12524/12600
12525/12600
12526/12600
12527/12600
12528/12600
12529/12600
12530/12600
12531/12600
12532/12600
12533/12600
12534/12600
12535/12600
12536/12600
12537/12600
12538/12600
12539/12600
12540/12600
12541/12600
12542/12600
12543/12600
12544/12600
12545/12600
12546/12600
ERROR ON 621707598
12547/12600
12548/12600
12549/12600
12550/12600
12551/12600
12552/12600
12553/12600
12554/12600
12555/12600
12556/12600
12557/12600
12558/12600
12559/12600
12560/12600
12561/12600
12562/12600
12563/12600
12564/12600
12565/12600
12566/12600
12567/12600
12568/12600
12569/12600
12570/12600
12571/12600
12572/12600
12573/12600
ERROR ON 4174445898
12574/12600
12575/12600
12576/12600
12577/12600
12578/12600
12579/12600
12580/12600
12581/12600
12582/12600
12583/12600
12584/12600
12585/12600
12586/12600
12587/12600
12588/12600
12589/12600
12590/12600
12591/12600
12592/12600
12593/12600
12594/12600
12595/12600
12596/12600
12597/12600
12598/12600
12599/12600
12600/12600


In [16]:
unique_nodes

{'597811237',
 '4212883643',
 '582701512',
 '1471342447',
 '5286036493',
 '3174700376',
 '5340504048',
 '310106320',
 '1675048509',
 '340065865',
 '253092251',
 '597811342',
 '599023989',
 '5234538634',
 '2296082203',
 '266197782',
 '269946533',
 '3090820936',
 '318536592',
 '253088268',
 '583412258',
 '1474530558',
 '1184831850',
 '595272009',
 '563021342',
 '2101684819',
 '331236097',
 '487097636',
 '253091860',
 '2226798',
 '1234078686',
 '3174694960',
 '580046878',
 '304577577',
 '5273357406',
 '573250813',
 '1831336402',
 '1789818946',
 '31296478',
 '1396425853',
 '2472637798',
 '3076686913',
 '1262394230',
 '4605018492',
 '4947137560',
 '561509451',
 '4220746399',
 '35296873',
 '1102248315',
 '3102621416',
 '1024140234',
 '3062959502',
 '470537572',
 '1030114452',
 '677515188',
 '598696907',
 '1475318854',
 '580046903',
 '673392338',
 '1240258069',
 '2071275464',
 '3409712653',
 '574869537',
 '330905280',
 '1022383029',
 '2132708774',
 '575172295',
 '253007707',
 '857558115',
 '3

In [15]:
unique_nodes.

AttributeError: 'set' object has no attribute 'get'

20: La nonna
21: Sukkho
22: Caffe Bellini
23: Etoile Blanche
24: Sakura
25: ?
26: ?
27: Aux fleurs de la terre
28: ?
29: Zooburger / Duo thai sushi
30: 


In [75]:
node_cord_df = pd.read_csv("node_id_to_coordinates.csv")
print(node_cord_df.head())

      node_id        lat       lon
0   598199831  46.513593  6.616881
1   570122472  46.513593  6.616881
2   662611324  46.513593  6.616881
3   412473064  46.513593  6.616881
4  4788039256  46.513593  6.616881


In [9]:
df.columns

Index(['plat', 'plng', 'dlat', 'dlng', 't', 'road'], dtype='object')

In [15]:
# how many unique road we have?
display(df.shape[0])
display(df.road.nunique())  # out of 2000 only 935 are unique

2000

935

In [ ]:
XXX to do XXX

## Count Nodes in a Road

In [16]:
# Analyse routes (list of nodes) taken be drivers
df['numNodes'] = df.road.apply(lambda x: len(x))
display(df.head())

plat      plng       dlat      dlng         t  \
0  46.517116  6.630342  46.508528  6.627598  11:15:00   
1  46.517116  6.630342  46.521429  6.656347  20:00:00   
2  46.517116  6.630342  46.558792  6.680104  18:15:00   
3  46.520649  6.635173  46.519590  6.634718  18:45:00   
4  46.510731  6.630553  46.515950  6.657525  19:00:00   

                                                road  numNodes  
0  [311463563, 2940544168, 311463564, 561463459, ...       683  
1  [311463563, 2940544168, 311463564, 561463459, ...      2638  
2  [311463563, 2940544168, 311463564, 561463459, ...      3949  
3  [2223091605, 3786724165, 3786724164, 302530573...      1401  
4  [3516999773, 3817926827, 567951313, 567951312,...      2224

In [17]:
# According to data for a delivery a driver passes avg of 2380 nodes.
df.numNodes.describe()

count    2000.000000
mean     2379.610000
std      1207.810329
min       135.000000
25%      1373.000000
50%      2126.000000
75%      3614.000000
max      7429.000000
Name: numNodes, dtype: float64

Max num of nodes in the road is 7429 whereas min is 159.

In [23]:
# check if in a node there is duplicated node
df['duplicated_nodes_in_road'] = df.road.apply(lambda x: len(x) == len(set(x)))
display(df[df.duplicated_nodes_in_road==True])  # There is none -- santiy check is correct
display(df.head())
df.drop(columns=['duplicated_nodes_in_road'], inplace=True)

Empty DataFrame
Columns: [plat, plng, dlat, dlng, t, road, numNodes, duplicated_nodes_in_road]
Index: []

plat      plng       dlat      dlng         t  \
0  46.517116  6.630342  46.508528  6.627598  11:15:00   
1  46.517116  6.630342  46.521429  6.656347  20:00:00   
2  46.517116  6.630342  46.558792  6.680104  18:15:00   
3  46.520649  6.635173  46.519590  6.634718  18:45:00   
4  46.510731  6.630553  46.515950  6.657525  19:00:00   

                                                road  numNodes  \
0  [311463563, 2940544168, 311463564, 561463459, ...       683   
1  [311463563, 2940544168, 311463564, 561463459, ...      2638   
2  [311463563, 2940544168, 311463564, 561463459, ...      3949   
3  [2223091605, 3786724165, 3786724164, 302530573...      1401   
4  [3516999773, 3817926827, 567951313, 567951312,...      2224   

   duplicated_nodes_in_road  
0                     False  
1                     False  
2                     False  
3                     False  
4                     False

In [32]:
df.dtypes

plat        float64
plng        float64
dlat        float64
dlng        float64
t            object
road         object
numNodes      int64
dtype: object

In [90]:
# change road type from string to list
from ast import literal_eval
df['road_list'] = df.road.apply(lambda x: literal_eval(x))
df.dtypes

plat         float64
plng         float64
dlat         float64
dlng         float64
t             object
road          object
numNodes       int64
road_list     object
dtype: object

In [92]:
df.head()
df.drop(columns=['road_list'])

plat      plng       dlat      dlng         t  \
0     46.517116  6.630342  46.508528  6.627598  11:15:00   
1     46.517116  6.630342  46.521429  6.656347  20:00:00   
2     46.517116  6.630342  46.558792  6.680104  18:15:00   
3     46.520649  6.635173  46.519590  6.634718  18:45:00   
4     46.510731  6.630553  46.515950  6.657525  19:00:00   
5     46.520649  6.635173  46.558792  6.680104  18:30:00   
6     46.517116  6.630342  46.537950  6.571898  18:00:00   
7     46.520649  6.635173  46.558792  6.680104  20:15:00   
8     46.518914  6.637912  46.537411  6.653249  19:30:00   
9     46.517116  6.630342  46.524409  6.629936  19:15:00   
10    46.520649  6.635173  46.530030  6.572823  19:45:00   
11    46.527252  6.623526  46.558792  6.680104  21:00:00   
12    46.517116  6.630342  46.509917  6.622856  18:15:00   
13    46.520649  6.635173  46.541796  6.623339  11:15:00   
14    46.517116  6.630342  46.515180  6.633784  18:45:00   
15    46.520649  6.635173  46.526995  6.622243  20:30:00   
16    46.531433  6.626430  46.558792  6.680104  19:15:00   
17    46.520649  6.635173  46.518409  6.640542  19:30:00   
18    46.527087  6.604758  46.517731  6.614065  18:30:00   
19    46.531433  6.626430  46.515158  6.614027  18:00:00   
20    46.510731  6.630553  46.520619  6.639434  18:45:00   
21    46.522014  6.627691  46.527379  6.570007  18:15:00   
22    46.517116  6.630342  46.516208  6.637349  19:00:00   
23    46.517671  6.638005  46.515158  6.614027  20:45:00   
24    46.520649  6.635173  46.508519  6.654169  19:30:00   
25    46.520649  6.635173  46.519211  6.615578  19:45:00   
26    46.517116  6.630342  46.520647  6.645823  19:00:00   
27    46.520649  6.635173  46.558792  6.680104  18:15:00   
28    46.520649  6.635173  46.558792  6.680104  19:45:00   
29    46.517116  6.630342  46.520394  6.630678  18:00:00   
...         ...       ...        ...       ...       ...   
1970  46.520649  6.635173  46.558792  6.680104  20:45:00   
1971  46.531433  6.626430  46.514832  6.650838  20:30:00   
1972  46.520649  6.635173  46.558792  6.680104  11:15:00   
1973  46.517116  6.630342  46.506952  6.628873  19:30:00   
1974  46.517116  6.630342  46.558792  6.680104  19:45:00   
1975  46.520649  6.635173  46.516796  6.625463  12:15:00   
1976  46.520649  6.635173  46.526487  6.634099  18:00:00   
1977  46.521701  6.639626  46.529810  6.632502  19:45:00   
1978  46.520649  6.635173  46.519235  6.614109  19:30:00   
1979  46.517116  6.630342  46.529772  6.636233  19:00:00   
1980  46.520649  6.635173  46.512453  6.635183  19:45:00   
1981  46.520649  6.635173  46.514442  6.641614  19:15:00   
1982  46.518914  6.637912  46.519590  6.634718  18:30:00   
1983  46.517116  6.630342  46.520228  6.639790  18:00:00   
1984  46.520649  6.635173  46.517307  6.630307  18:30:00   
1985  46.520649  6.635173  46.513165  6.624953  19:45:00   
1986  46.520649  6.635173  46.558792  6.680104  17:45:00   
1987  46.517116  6.630342  46.508295  6.646022  20:00:00   
1988  46.520649  6.635173  46.526587  6.645086  20:45:00   
1989  46.522426  6.642152  46.516208  6.637349  19:00:00   
1990  46.520649  6.635173  46.513999  6.635531  19:45:00   
1991  46.520649  6.635173  46.558792  6.680104  19:15:00   
1992  46.523953  6.625008  46.530030  6.572823  18:30:00   
1993  46.520649  6.635173  46.558792  6.680104  20:30:00   
1994  46.522014  6.627691  46.519590  6.634718  19:30:00   
1995  46.517116  6.630342  46.527403  6.615034  18:15:00   
1996  46.520649  6.635173  46.558792  6.680104  18:15:00   
1997  46.523815  6.636847  46.514855  6.639087  19:30:00   
1998  46.520649  6.635173  46.558792  6.680104  19:00:00   
1999  46.520946  6.633680  46.513892  6.615793  21:00:00   

                                                   road  numNodes  
0     [311463563, 2940544168, 311463564, 561463459, ...       683  
1     [311463563, 2940544168, 311463564, 561463459, ...      2638  
2     [311463563, 2940544168, 311463564, 561463459, ...      3949  
3     [2223091605,

In [93]:
df.to_csv('dataviz_numNodes.csv')

In [83]:
# given a node find the all routes using this node
# which deliveries use this node? return the ones id
def findRoutes(node):
    #print('Node id: ',node)
    id_deliveries = []
    for idx, d in df.iterrows() :
        if(node in d.road_list):
            id_deliveries.append(idx)
    return  id_deliveries

In [84]:
node_id = 311463563
res = findRoutes(node_id)
print('Node ',node_id, ' used ', len(res),' different deliveries.')
print('The deliveries id are: ',res)

Node  311463563  used  405  different deliveries.
The deliveries id are:  [0, 1, 2, 6, 9, 12, 14, 22, 26, 29, 41, 43, 46, 49, 50, 54, 60, 65, 68, 79, 88, 89, 100, 103, 105, 108, 116, 119, 127, 133, 157, 158, 163, 173, 174, 178, 181, 186, 190, 194, 195, 206, 213, 217, 225, 230, 232, 236, 246, 247, 252, 257, 259, 278, 292, 303, 308, 312, 315, 322, 330, 331, 334, 347, 350, 364, 377, 378, 390, 396, 402, 404, 413, 420, 424, 427, 430, 431, 432, 436, 446, 455, 456, 457, 459, 461, 475, 484, 487, 489, 490, 491, 493, 496, 497, 499, 502, 508, 512, 514, 518, 529, 530, 535, 538, 548, 549, 556, 573, 574, 577, 579, 587, 589, 609, 610, 616, 618, 623, 629, 636, 639, 643, 644, 660, 667, 671, 679, 683, 687, 688, 689, 693, 707, 725, 726, 727, 730, 737, 738, 740, 741, 746, 751, 753, 754, 757, 761, 763, 765, 770, 780, 783, 784, 792, 804, 815, 816, 818, 841, 844, 846, 848, 856, 861, 862, 870, 874, 875, 884, 886, 887, 888, 891, 897, 898, 899, 904, 907, 908, 913, 914, 916, 927, 930, 933, 936, 940, 946, 950, 95

In [85]:
nodes = node_cord_df.copy() 
nodes.head()

node_id        lat       lon
0   598199831  46.513593  6.616881
1   570122472  46.513593  6.616881
2   662611324  46.513593  6.616881
3   412473064  46.513593  6.616881
4  4788039256  46.513593  6.616881

In [86]:
# takes some time be patient or read the below file instead reexecuting - 'nodes_usage_delivery_ids.csv'
# Do this for all nodes and store the results in a nodes df
nodes['delivery_ids'] = nodes.node_id.apply(lambda x: findRoutes(x))
nodes['tot_usage'] = nodes.delivery_ids.apply(lambda x: len(x))

In [87]:
nodes.head()

node_id        lat       lon  \
0       598199831  46.513593  6.616881   
1       570122472  46.513593  6.616881   
2       662611324  46.513593  6.616881   
3       412473064  46.513593  6.616881   
4      4788039256  46.513593  6.616881   
5      3076786032  46.513593  6.616881   
6      2789084918  46.513593  6.616881   
7       275366462  46.513593  6.616881   
8       269236712  46.513593  6.616881   
9       573250837  46.513593  6.616881   
10     2727456548  46.513593  6.616881   
11      583408360  46.513593  6.616881   
12     4200605895  46.513593  6.616881   
13     4212883645  46.513593  6.616881   
14      598991447  46.513593  6.616881   
15      861385299  46.513593  6.616881   
16      567927282  46.513593  6.616881   
17      616680119  46.513593  6.616881   
18      280798828  46.513593  6.616881   
19     1678434752  46.513593  6.616881   
20      567496976  46.513593  6.616881   
21     1242522851  46.513593  6.616881   
22     1184824117  46.513593  6.616881   
23     1242524205  46.513593  6.616881   
24      580046802  46.513593  6.616881   
25       60437926  46.513593  6.616881   
26     1511515191  46.513593  6.616881   
27      256188252  46.513593  6.616881   
28     1025658621  46.513593  6.616881   
29      260928348  46.513593  6.616881   
...           ...        ...       ...   
12570   309457748  46.513593  6.616881   
12571   565051471  46.513593  6.616881   
12572   304577576  46.513593  6.616881   
12573  3042147382  46.513593  6.616881   
12574   659134576  46.513593  6.616881   
12575   289475203  46.513593  6.616881   
12576   609224230  46.513593  6.616881   
12577   563021456  46.513593  6.616881   
12578   570083978  46.513593  6.616881   
12579   330902469  46.513593  6.616881   
12580   258755820  46.513593  6.616881   
12581   318524633  46.513593  6.616881   
12582   313711863  46.513593  6.616881   
12583    35297473  46.513593  6.616881   
12584   583408395  46.513593  6.616881   
12585   272832470  46.513593  6.616881   
12586   313700342  46.513593  6.616881   
12587   420770598  46.513593  6.616881   
12588   580065632  46.513593  6.616881   
12589  5331033803  46.513593  6.616881   
12590  5337890157  46.513593  6.616881   
12591  1474513527  46.513593  6.616881   
12592   332269630  46.513593  6.616881   
12593  1080666649  46.513593  6.616881   
12594  4472670146  46.513593  6.616881   
12595  1181229532  46.513593  6.616881   
12596  2313165463  46.513593  6.616881   
12597  5280752785  46.513593  6.616881   
12598   427959536  46.513593  6.616881   
12599   983813495  46.513593  6.616881   

                                            delivery_ids  tot_usage  
0      [257, 389, 428, 472, 1328, 1607, 1617, 1710, 1...         11  
1                                            [243, 1293]          2  
2                            [677, 747, 764, 1193, 1823]          5  
3                                                 [1097]          1  
4                                                 [1894]          1  
5      [1, 2, 3, 8, 17, 24, 26, 31, 41, 43, 44, 46, 4...        467  
6                             [111, 219, 402, 623, 1394]          5  
7                                                  [516]          1  
8                                        [148, 248, 670]          3  
9                          [428, 1497, 1607, 1828, 1857]          5  
10                     [209, 479, 522, 1526, 1570, 1862]          6  
11     [39, 143, 276, 296, 300, 341, 354, 376, 409, 4...         46  
12                                                 [408]          1  
13     [1, 109, 184, 221, 234, 270, 305, 339, 381, 52...         29  
14                                       [194, 274, 656]          3  
15     [423, 432, 605, 644, 653, 859, 1097, 1292, 140...         13  
16     [32, 104, 182, 189, 265, 284, 372, 406, 414, 5...         22  
17                            [289, 292, 601, 674, 1668]          5  
18        [410, 994, 1041, 1201, 1363, 1567, 1700, 1931]          8  
19   

In [88]:
nodes.to_csv('nodes_usage_delivery_ids.csv')